### DISCUSSION TITLE
select.select is always blocked.

### DISCUSSION THREAD

**Discussion by shechren** [Rapptz/discord.py#9112] [Q&A] [answered]
let me excuse my poor Eng Skills.

i tried the socket network. but it's so inefficient. however i found the select.select instead.
but i received the result of Error about **Shard ID None heartbeat blocked**
so i tried again that i added setblocking(False) for my socket and descendant socket twice.
and i could see the result "False" from print(getblocking())
nevertheless heartbeat blocked is always coming even when i didn't connect to any connection to server.
please help me what should i do.
i write my code below this.

```
import asyncio
import discord
from discord.ext import commands
import os
import select
import socket

class SocTCP(commands.Cog):
    def __init__(self, bot):
        self.bot = bot
        self.HOST = "myHOST"
        self.PORT = int(os.environ.get("PORT", "myPORT"))

    async def connect(self):
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.bind(("", self.PORT))
            s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            s.listen()
            print("server connected")
            readsocks = [s]
            channel = self.bot.get_channel("my Channel id")
            s.setblocking(False)
            while True:

                try:
                    readables, writeables, exceptions = select.select([s], [], [], 0)
                    for data in readables:
                        if not (readables, writeables, exceptions):
                            await asyncio.sleep(1)
                        if data == s:
                            conn, addr = s.accept()
                            conn.setblocking(False)
                            readsocks.append(conn)
                            readsocks = list(set(readsocks))
                            print(readsocks)
                            print(f"connected new friend! {addr}")
                        else:
                            message = data.recv(1024).decode()
                            print(f"{message}")
                            await channel.send(f"{message}")
                except BlockingIOError:
                    print("block")
                    await asyncio.sleep(1)
                except discord.errors.HTTPException:
                    readsocks.remove(conn)
                    conn.close()
                    print(f"client closed")


async def setup(bot):
    await bot.add_cog(SocTCP(bot))
```

**Comment by LostLuma** [comment, marked as answer]
Hello, this seems to be a misunderstanding of what the `socket.setblocking` method on a socket does.
You can view the documentation of the effects of non-blocking mode in the [Python documentation here](https://docs.python.org/3/library/socket.html#socket-timeouts).

If you'd like to connect to a socket in an async enviroment you should use [`Loop.create_connection`](https://docs.python.org/3/library/asyncio-eventloop.html#opening-network-connections) instead, you can get the active loop using `asyncio.get_running_loop()`.

**Reply by shechren** [reply]
thx for your answer. i used this for connect in main.py
`from sv.socTCP import SocTCP
        constclass = SocTCP(bot)
        t = asyncio.create_task(constclass.connect())
        await t`
i wanna ask more. what is the difference between make a new loop and create task?

**Reply by LostLuma** [reply]
Creating a new loop would replace the existing event loop that already exists from running your bot, so it would not help you with your issue here. A task simply schedules an async function to run on the existing event loop, which you're free to do at any time to run things in the background.

**Reply by shechren** [reply]
really thanks for your kind. Sincerely
